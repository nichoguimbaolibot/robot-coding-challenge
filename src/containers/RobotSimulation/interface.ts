export interface IRobot {
  isPosition: boolean;
  recordedMovements: IRecordedMovements[];
  robotDirection: number;
  robotXAxisPlacement: number;
  robotYAxisPlacement: number;
  type: string;
}

export interface IRecordedMovements {
  robotDirection: number;
  id: number;
  robotXAxisPlacement: number;
  time: Date;
  robotYAxisPlacement: number;
  type: string;
}
