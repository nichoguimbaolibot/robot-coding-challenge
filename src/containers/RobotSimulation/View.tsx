import React, { FunctionComponent, useState } from 'react';
import { Howl } from 'howler';
import Notification from 'antd/lib/notification';
import MESSAGES from 'constant/message';
import {
  isCardinalDirectionNorth,
  isCardinalDirectionEast,
  isCardinalDirectionSouth,
  isCardinalDirectionWest,
} from 'utils/validatePosition';
import { NORTH, EAST, SOUTH, WEST } from 'constant/cardinalDirection';
import { generateUniqueId } from 'utils/generateUniqueId';
import Controls from './components/Controls';
import RobotPlacement from './components/RobotPlacement';
import Robot from './components/Robot';
import Table from './components/Table';

import { MAX_BOXES } from 'constant/app';

const RobotCodingChallenge: FunctionComponent = (): JSX.Element => {
  const [robotXAxisPlacement, setRobotXAxisPlacement] = useState(0);
  const [robotYAxisPlacement, setRobotYAxisPlacement] = useState(0);
  const [robotDirection, setRobotDirection] = useState(0);
  const [report, setReport] = useState({
    robotDirection: 0,
    id: generateUniqueId(),
    robotXAxisPlacement: 0,
    robotYAxisPlacement: 0,
  });
  const [isPosition, setPosition] = useState(false);

  const handleRobotPlacement = (
    xAxis: number,
    yAxis: number,
    cardinalDirection: number,
  ): void => {
    if (xAxis < 0 || xAxis > 4 || yAxis < 0 || yAxis > 4) {
      showErrorNotificationMessage(MESSAGES.INVALID_POSITION);
      return;
    } else {
      robotPlacementSounds();
      setPosition(true);
      setRobotDirection(cardinalDirection);
      setRobotXAxisPlacement(xAxis);
      setRobotYAxisPlacement(yAxis);
      handleUpdateReport(cardinalDirection, yAxis, xAxis);
    }
  };

  const handleRobotMovement = (): void => {
    if (!isPosition) {
      showErrorNotificationMessage(MESSAGES.ROBOT_NOT_FOUND);
      return;
    }
    const currentDirection = robotDirection % 360;
    let xAxisPlacement = robotXAxisPlacement;
    let yAxisPlacement = robotYAxisPlacement;
    if (isCardinalDirectionNorth(currentDirection)) {
      yAxisPlacement = yAxisPlacement + 1;
    }

    if (isCardinalDirectionEast(currentDirection)) {
      xAxisPlacement = xAxisPlacement + 1;
    }

    if (isCardinalDirectionSouth(currentDirection)) {
      yAxisPlacement = yAxisPlacement - 1;
    }

    if (isCardinalDirectionWest(currentDirection)) {
      xAxisPlacement = xAxisPlacement - 1;
    }

    if (
      xAxisPlacement < 0 ||
      xAxisPlacement > 4 ||
      yAxisPlacement < 0 ||
      yAxisPlacement > 4
    ) {
      showErrorNotificationMessage(MESSAGES.MOVEMENT_NOT_ALLOWED);
      return;
    } else {
      robotMovementSound();
      setRobotXAxisPlacement(xAxisPlacement);
      setRobotYAxisPlacement(yAxisPlacement);
      handleUpdateReport(currentDirection, yAxisPlacement, xAxisPlacement);
    }
  };

  const handleChangeMovementToLeft = (): void => {
    if (!isPosition) {
      showErrorNotificationMessage(MESSAGES.ROBOT_NOT_FOUND);
      return;
    }
    const nextDirection = robotDirection - 90;
    robotRotateSound();
    setRobotDirection(nextDirection);
  };

  const handleChangeMovementToRight = (): void => {
    if (!isPosition) {
      showErrorNotificationMessage(MESSAGES.ROBOT_NOT_FOUND);
      return;
    }
    const nextDirection = robotDirection + 90;
    robotRotateSound();
    setRobotDirection(nextDirection);
  };

  const validateDirection = (position: number): string => {
    if (isCardinalDirectionNorth(position)) {
      return NORTH;
    } else if (isCardinalDirectionEast(position)) {
      return EAST;
    } else if (isCardinalDirectionSouth(position)) {
      return SOUTH;
    } else if (isCardinalDirectionWest(position)) {
      return WEST;
    }

    return 'INVALID FACE POSITION';
  };

  const handleUpdateReport = (
    currentDirection: number,
    yAxis: number,
    xAxis: number,
  ): void => {
    const currentReport = {
      robotDirection: currentDirection,
      id: generateUniqueId(),
      robotXAxisPlacement: xAxis,
      robotYAxisPlacement: yAxis,
    };
    setReport(currentReport);
  };

  const showErrorNotificationMessage = (message: string): void => {
    Notification.error({
      placement: 'bottomRight',
      message: 'Err...',
      description: message,
    });
    errorNoSound();
  };

  const handleReportRobotPosition = (): void => {
    if (!isPosition) {
      Notification.success({
        placement: 'bottomRight',
        message: 'Report Position',
        description: MESSAGES.ROBOT_NOT_FOUND,
      });
      return;
    }
    const message = `X-Axis: ${report.robotXAxisPlacement} Y:Axis: ${
      report.robotYAxisPlacement
    } Cardinal Direction: ${validateDirection(report.robotDirection)}`;
    Notification.success({
      placement: 'bottomRight',
      message: 'Report Position',
      description: message,
    });
  };

  const robotRotateSound = (): void => {
    const robotRotateSound = new Howl({
      src: ['/robot-rotate.mp3'],
    });
    robotRotateSound.play();
  };

  const robotMovementSound = (): void => {
    const robotMovementSound = new Howl({
      src: ['/robot-moving.mp3'],
    });
    robotMovementSound.play();
  };

  const errorNoSound = (): void => {
    const errorNoSound = new Howl({
      src: ['/error-no.mp3'],
    });
    errorNoSound.play();
  };

  const robotPlacementSounds = (): void => {
    const robotPlacementSound = new Howl({
      src: ['/robot-placement.mp3'],
    });
    robotPlacementSound.play();
  };

  return (
    <div className="robot-simulation-section">
      <div className="table-container">
        <Table rowBoxes={MAX_BOXES} />

        <Robot
          xAxisPlacement={robotXAxisPlacement}
          yAxisPlacement={robotYAxisPlacement}
          robotDirection={robotDirection}
          isPosition={isPosition}
        />
      </div>
      <div className="commands">
        <Controls
          onReportRobotPosition={handleReportRobotPosition}
          onChangeRobotMovement={handleRobotMovement}
          onChangeMovementToLeft={handleChangeMovementToLeft}
          onChangeMovementToRight={handleChangeMovementToRight}
        />
        <RobotPlacement onRobotPlacement={handleRobotPlacement} />
      </div>
    </div>
  );
};

export default RobotCodingChallenge;
