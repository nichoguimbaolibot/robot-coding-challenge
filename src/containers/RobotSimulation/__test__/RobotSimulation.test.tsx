import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import RobotSimulation from 'containers/RobotSimulation';

afterEach(cleanup);

describe('test RobotSimulation Module', () => {
  it('renders ROBOTSIMULATION component', () => {
    const { asFragment } = render(<RobotSimulation />);
    expect(asFragment()).toMatchSnapshot();
  });
});
