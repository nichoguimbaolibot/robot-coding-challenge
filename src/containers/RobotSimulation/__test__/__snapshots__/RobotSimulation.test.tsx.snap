// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test RobotSimulation Module renders ROBOTSIMULATION component 1`] = `
<DocumentFragment>
  <div
    class="robot-simulation-section"
  >
    <div
      class="table-container"
    >
      <div
        class="table-box-section"
      >
        <div
          class="table-item"
        >
          <div
            class="vertical-coordinates"
          >
            <section
              class="coordinate-section -block"
            >
              <span
                class="coordinate-label"
              >
                0
              </span>
              <span
                class="coordinate-label"
              >
                1
              </span>
              <span
                class="coordinate-label"
              >
                2
              </span>
              <span
                class="coordinate-label"
              >
                3
              </span>
              <span
                class="coordinate-label"
              >
                4
              </span>
            </section>
          </div>
          <div
            class="box-generator"
          >
            <div
              class="box-container"
              data-testid="box-container-test"
            >
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
            </div>
            <div
              class="box-container"
              data-testid="box-container-test"
            >
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
            </div>
            <div
              class="box-container"
              data-testid="box-container-test"
            >
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
            </div>
            <div
              class="box-container"
              data-testid="box-container-test"
            >
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
            </div>
            <div
              class="box-container"
              data-testid="box-container-test"
            >
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
              <div
                class="box"
                data-testid="box-test"
              />
            </div>
            <section
              class="coordinate-section -inline"
            >
              <span
                class="coordinate-label"
              >
                0
              </span>
              <span
                class="coordinate-label"
              >
                1
              </span>
              <span
                class="coordinate-label"
              >
                2
              </span>
              <span
                class="coordinate-label"
              >
                3
              </span>
              <span
                class="coordinate-label"
              >
                4
              </span>
            </section>
          </div>
        </div>
      </div>
      <section
        class="robot-section"
        data-testid="robot-test"
        id="robot"
        style="display: none;"
      />
    </div>
    <div
      class="commands"
    >
      <div
        class="direction-section"
      >
        <h4
          class="title"
        >
          Movement and Positions
        </h4>
        <button
          class="ant-btn left-button ui-button-primary"
          data-testid="button-move-left-test"
          type="button"
        >
          <span>
            LEFT
          </span>
        </button>
        <button
          class="ant-btn right-button ui-button-primary"
          data-testid="button-move-right-test"
          type="button"
        >
          <span>
            RIGHT
          </span>
        </button>
        <button
          class="ant-btn move-button ui-button-alpha"
          data-testid="button-move-test"
          type="button"
        >
          <span>
            MOVE
          </span>
        </button>
        <button
          class="ant-btn"
          data-testid="button-report-test"
          type="button"
        >
          <span>
            REPORT
          </span>
        </button>
      </div>
      <div
        class="robot-placement-section"
      >
        <h4>
          Robot Placement
        </h4>
        <input
          class="ant-input input-xAxis"
          data-testid="x-axis-test"
          id="input-xAxis"
          placeholder="x-axis position"
          type="text"
          value=""
        />
        <input
          class="ant-input input-yAxis"
          data-testid="y-axis-test"
          id="input-yAxis"
          placeholder="y-axis position"
          type="text"
          value=""
        />
        <div
          class="select-position ant-select ant-select-enabled"
          id="select-position"
        >
          <div
            aria-autocomplete="list"
            aria-controls="test-uuid"
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-select-selection
            ant-select-selection--single"
            data-testid="direction-test"
            role="combobox"
            tabindex="0"
          >
            <div
              class="ant-select-selection__rendered"
            >
              <div
                class="ant-select-selection__placeholder"
                style="display: block; user-select: none;"
                unselectable="on"
              >
                ex. North
              </div>
            </div>
            <span
              class="ant-select-arrow"
              style="user-select: none;"
              unselectable="on"
            >
              <i
                aria-label="icon: down"
                class="anticon anticon-down ant-select-arrow-icon"
              >
                <svg
                  aria-hidden="true"
                  class=""
                  data-icon="down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                  />
                </svg>
              </i>
            </span>
          </div>
        </div>
        <button
          class="ant-btn place-robot-btn ant-btn-primary"
          data-testid="place-robot-test"
          type="button"
        >
          <span>
            PLACE
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
